### ISSUE TITLE
1837 improvements and bugs (2)

### ISSUE THREAD

**Issue by erik-vos** [Rails-18xx/Rails#341]
See #329 for the previous list, of which 3 topics have been moved to here.
The low priority feature requests [3,7,13] of that issue have not been copied here.
@neutronc is working on a background map (topics [8,14]).

(Author of issues [1-3]: John David Galt)
[1, was 2] The use of multiple "start rounds" is a choice I made when
I worked on it, but I now believe it was a mistake.
Technically the only the first screen of that type is the
"start packet," defined in the rules as "it all has to sell
out before an operating round can take place."

The way I should have handled the sale of the K1-K3 and U1-U3
is to make those the first companies that are available, using
the existing mechanism of "layers" that limits the order of
starting new companies in 1829, 1835, and 1853.
**Status**: done.

[2, was 5] Bug: In my test game today, all of the K1-K3 sold out before
any operating round took place, but of the U1-U3, only the U1
president's share sold. We played one operating round. When it
ended the "start round" window displayed again, to sell the rest
of U1-U3; but only one player had enough money to buy any of the
shares. That player bought one, and the other two players passed
-- and then Rails ended the round before the player with money
could buy another share!
Save file: 1837_20210113_2138_OperatingRound 0.1.rails
**Status**: probably related to [1]. Not retested, saves file is obsolete.

[3, was 6] Worse bug: The "forced undo" mechanism would NOT take me back
into that "start round" to let the player buy more shares. When
I tried it, the "Leave history" and "Play from here" buttons
spontaneously disappeared from the Game Report window.
**Status**: probably related to [1]. Not retested.

**Comment by erik-vos** [comment]
An already closed 1837 issue has been reopened:

[4, was 18] Wien (Vienna) cannot be upgraded to green. The connection is seen, but the upgrade is disallowed (pink border).
**Original status**: This was a hard one. It turned out, that `TileUpgrade.assignStations()` could not deal with the six-fold symmetry of tile 427: all rotations of this tile are equivalent and thus valid, but that method found none, and thus no stationmapping! As I don't understand what that method is doing, I had to add my own code to fix it in the obvious way: for each side, map the old and new stations having tracks to that side. I have also implemented the option to prescribe a fixed tile orientation, and used it for tiles 427 and 436 (the latter would need it anyway).
**New status**: I had to reopen this problem. As described in the Game Notes: the tokens in the green Vienna tile had to be relaid manually (i.e. by answering a prompt for each token). That is rather unhelpful.
I have now found two separate causes for this behaviour, which I will describe separately.

[4.1] The standard station mapping code demands that, during the initial setup of the `TileUpgrade` objects, no unmapped stations remain on both the old and the new tile. If not, no station mapping is created at all. This explains why the code could not find any valid orientation for the green tile on Vienna, as the old and new tiles have 4 and 6 stations, respectively. Effectively, the code does not allow new unconnected stations to be added during an upgrade.
I see two possible solutions:
a) drop the requirement that the new tile may not have unmappable new stations. 
b) make this requirement optional, in that it must be configured to be applied. 
**Question**: is there any need to apply this check at all, so we need b)? I don't see one, for now I think that a) is sufficient.

I have applied a), and debugging now shows correct station mapping. Nevertheless, the manual token relay prompts did not disappear yet.

[4.2] For some time I had wondered what the reason was of having the `relayBaseTokens` attribute of the `Upgrade` tags in `TileSet.xml`. Should tokens not always be relaid?  I now understand that its function was to enforce _manual_ token relaying in an upgrade, apparently circumventing earlier Vienna upgrade problems. So it's an emergency measure only.
Removing this attribute from the Vienna yellow-to-green upgrade now lets the automatic relay be effective. Fixed!

I have also seen another upgrade doing the same, I will check that one too.

**Comment by erik-vos** [comment]
Also implemented:

[5, was 3] Multi-column display of start packets. Applied to the 1835 and 1837 base games.

[6, new] Added `displayName` attribute to the `<StartItem>` tag, so items can (optionally) be named separately from their contents. Applied to the 1835 and 1837 "double" items: `"NF+By"` and `"SmB+S1"`.

Will now look at [1].

**Comment by erik-vos** [comment]
[1] Whilst converting the 1837 subsequent start rounds to the stock round(s), it turned out that I had to take a fresh look on availability (i.e. buyability in a stock round). This attribute could be defined both on company and on certificate level.

On _company_ level: an unavailable company must explicitly be released (i.e. made available). In 1835 and 1837, this is done by fairly extensive code in methods overriding StockRound.gameSpecificChecks().
On _certificate_ level: unavailable certificates are typically used to exchange privates or minors.
(Note: some certs were marked "reserved", but that attribute was not even parsed).

To make this all working in a practical way, I have (re)defined the use of availability on both levels as follows:

1. During initial setup:
If a company is marked unavailable, all its certificates go to the 'unavailable' Bank portfolio.
If a company is marked available, only the certificates marked available go to the IPO, the remainder goes to 'unavailable'.
(BTW I had great difficulty to find where the certificates were moved to these Bank portfolios. It took me more than a day to discover that this happened in the class where I least expected it: the Bank!  I now understand why.)

2. When a company is released, only the certificates marked available go to the IPO.

I'm also simplifying the aforementioned code, with the ultimate goal to prepare making all company release rules configurable in a next development step.

**Comment by erik-vos** [comment]
[1]  1837 minor acquisition has now been moved to the StockRound (except the S minors, of course). Still some testing and minor bug fixing to do.  NOTE: this change will invalidate ALL previously saved files that have progressed beyond the initial purchases.

The original code for version 2 appeared to be based upon the misconception that passing is not allowed in the start packet auction. The rules (English version 2.0 of 2015 by Lonny, chapter I) say otherwise (my italicisations):
     - "[A player] may either start an auction (...) _or pass_."
     - "If the Start Packet is not completely sold out when the First Stock Round [= StartRound] ends [_because all players pass_], place any unsold items in a group with the two secondary (non-director) certificates of Ug minors #1 and #3." [Although it is not stated explicitly, I assume an operating round follows.]
     - "Any remaining Start Packet items are sold _at face value_" [in a stock round-like fashion, i.e., no more bidding].
 
For practical reasons, I have left acquisition of any remaining start packet items in a  start round that only allows direct buying. Because (unfortunately) the buy/bid settings are hardcoded in the start round constructor, I needed yet another subclass, which is named `StartRound_1837_2ndEd_buying`.

The kk and Ug start packets and their associated classes have been removed. `StartRound_1837_Coal` is renamed to `StartRound_1837_v1`.

**Comment by erik-vos** [comment]
Whilst working on the 1837 gradual company release scheme, I decided to bite the bullet and make this aspect completely configurable. Where applicable, `CompanyManager.xml` can now contain any number of `<ReleaseRule>` tags, which the program checks off in the specified sequence. For each rule, an instance is created of the new ReleaseRule class, which does the checking and releasing. For instance:
- in 1837v1: `<ReleaseRule sold="K1,K2,K3" released="U1,U2,U3"/>` makes the U minors buyable as soon as the K minors are sold.
- in 1835: `<ReleaseRule sold="BA:20" released="PR"/>` releases the 40% buyable Prussian certificates as soon as the Badische has started by buying its 20% presidency. This case critically depends on the PR being set to "unavailable" on company level as well as the remaining reserved 60% certs on certificate level.

This new feature has been applied to 1835 and 1837; I will look at 1825 later.

The next PR will follow when I have got a few test files, to protect all changes to 1837 thus far. To make that possible, I am currently trying to fix some issues with coal company merging. I will leave minor merging for later. Next thing will be writing a wiki page about all things related to certificates and shares.
